<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LMD Tutorial • LocalizedMarkerDetector</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LMD Tutorial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LocalizedMarkerDetector</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/LMD_demo.html">Example on mouse bone marrow scRNA-seq</a>
    </li>
    <li>
      <a href="../articles/LMD_cross_comparison_demo.html">Application to cross-sample comparison</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KlugerLab/LocalizedMarkerDetector">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LMD Tutorial</h1>
                        <h4 data-toc-skip class="author">Ruiqi Li,
Francesco Strino, Rihao Qu</h4>
            
            <h4 data-toc-skip class="date">08/20/2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KlugerLab/LocalizedMarkerDetector/blob/HEAD/vignettes/LMD_demo.Rmd"><code>vignettes/LMD_demo.Rmd</code></a></small>
      <div class="hidden name"><code>LMD_demo.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Localized Marker Detector (LMD) is a computational framework designed
for the identification of gene expression markers localized to specific
cell populations within single-cell RNA sequencing data. The major
workflow of LMD comprises the following three main steps:</p>
<ul>
<li>Step1. Constructing a cell-cell affinity graph</li>
<li>Step2. Diffusing the gene expression value across the cell
graph</li>
<li>Step3. Assigning a score to each gene based on the dynamics of its
diffusion process</li>
<li>Optional Downstream tasks
<ul>
<li><a href="#identifying-gene-modules">Identifying gene modules and
characterizing functional cell groups</a></li>
<li><a href="LMD_cross_comparison_demo.html">Cross-sample
comparison</a></li>
</ul>
</li>
</ul>
<p><img src="../reference/figures/LMD_workflow.png" width="100%" height="auto"></p>
</div>
<div class="section level2">
<h2 id="loading-required-r-packages">Loading required R packages<a class="anchor" aria-label="anchor" href="#loading-required-r-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##### Load required R libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://KlugerLab.github.io/LocalizedMarkerDetector/" class="external-link">"LocalizedMarkerDetector"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://satijalab.org/seurat" class="external-link">"Seurat"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SeuratWrappers"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/jabiru/tictoc" class="external-link">"tictoc"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preparing-input-data">Preparing input data<a class="anchor" aria-label="anchor" href="#preparing-input-data"></a>
</h2>
<div class="section level3">
<h3 id="loading-example-data">Loading example data<a class="anchor" aria-label="anchor" href="#loading-example-data"></a>
</h3>
<p>The preprocessed Seurat object for this tutorial can be downloaded
from <a href="https://figshare.com/ndownloader/files/13092380" class="external-link">figshare</a>. The
preprocessing step can be refer to the <a href="https://github.com/czbiohub-sf/tabula-muris/blob/master/00_data_ingest/All_FACS_Notebook.Rmd" class="external-link">Tabula
Muris workflow</a><span class="citation">(Consortium et al.
2018)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir.path0</span> <span class="op">=</span> <span class="st">"~/"</span> <span class="co"># Specify the directory path where you want to save the data</span></span>
<span><span class="va">file_name</span> <span class="op">=</span> <span class="st">"marrow_facs.rds"</span> <span class="co"># Specify the file name for saving the data</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir.path0</span>,<span class="va">file_name</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">6000</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://figshare.com/ndownloader/files/13092380"</span>, destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir.path0</span>,<span class="va">file_name</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">'libcurl'</span><span class="op">)</span></span>
<span>    <span class="co"># Check and upgrade the Seurat object if needed</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir.path0</span>,<span class="va">file_name</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">tiss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">tiss</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir.path0</span>,<span class="va">file_name</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">tiss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir.path0</span>,<span class="va">file_name</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extracting-input-data">Extracting input data<a class="anchor" aria-label="anchor" href="#extracting-input-data"></a>
</h3>
<p>Next, we prepare the following objects as input data from this Seurat
object.</p>
<ul>
<li><p><code>feature_space</code>: A matrix containing the first 20
principal components (PCs) from PCA.</p></li>
<li><p><code>visual_space</code>: A data frame containing the 2D t-SNE
coordinates</p></li>
<li><p><code>dat</code>: A matrix of log-normalized gene expression
values, where rows correspond to genes and columns correspond to
cells.</p></li>
<li><p><code>cell_label</code>: Metadata related to the cells, such as
cell type annotations, for visualization.</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="va">n_dim</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">feature_space</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">n_dim</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">visual_space</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">tsne</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">cell_label</span> <span class="op">=</span> <span class="va">tiss</span><span class="op">$</span><span class="va">cell_ontology_class</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="processing-input-data">Processing input data<a class="anchor" aria-label="anchor" href="#processing-input-data"></a>
</h3>
<p>We process the <code>dat</code> matrix by retaining only the genes
that are detected in more than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>10</mn><annotation encoding="application/x-tex">10</annotation></semantics></math>
cells and less than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>50</mn><mi>%</mi></mrow><annotation encoding="application/x-tex">50\%</annotation></semantics></math>
of cells. A gene is considered “detected” in a cell if its expression
level is greater than the median expression level of that cell.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Gene_detected_count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">dat</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">dat</span>,<span class="fl">2</span>,<span class="va">median</span><span class="op">)</span>,<span class="fl">1</span>,<span class="va">sum</span><span class="op">)</span></span>
<span><span class="va">selected_genes</span> <span class="op">=</span> <span class="op">(</span><span class="va">Gene_detected_count</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">Gene_detected_count</span> <span class="op">&lt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span> <span class="op">*</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">=</span> <span class="va">dat</span><span class="op">[</span><span class="va">selected_genes</span>,,drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-lmd-step-by-step">Running LMD Step by Step<a class="anchor" aria-label="anchor" href="#running-lmd-step-by-step"></a>
</h2>
<p>This provides a step-by-step tutorial for a better understanding of
LMD. To directly obtain the output of LMD, skip to <a href="#running-lmd-in-one-step">Running LMD in One Step</a>.</p>
<div class="section level3">
<h3 id="step1-constructing-a-cell-cell-affinity-graph">Step1: Constructing a cell-cell affinity graph<a class="anchor" aria-label="anchor" href="#step1-constructing-a-cell-cell-affinity-graph"></a>
</h3>
<p>We construct the cell-cell kNN graph (<code>K</code>= 5 in this
example) using <code>ConstructKnnGraph</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Construct knn graph</span></span>
<span><span class="va">knn_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/ConstructKnnGraph.html">ConstructKnnGraph</a></span><span class="op">(</span>knn <span class="op">=</span> <span class="fl">5</span>, feature_space <span class="op">=</span> <span class="va">feature_space</span><span class="op">)</span></span>
<span><span class="co">#&gt; Constructing KNN graph</span></span>
<span><span class="va">A</span> <span class="op">=</span> <span class="va">knn_result</span><span class="op">$</span><span class="va">adj_matrix</span> <span class="co"># Adjacency Matrix</span></span>
<span><span class="va">W</span> <span class="op">=</span> <span class="va">knn_result</span><span class="op">$</span><span class="va">graph</span> <span class="co"># Symmetrized Graph ((A + AT) / 2)</span></span>
<span></span>
<span><span class="co"># Plot knn graph</span></span>
<span><span class="fu"><a href="../reference/VisualizeGraph.html">VisualizeGraph</a></span><span class="op">(</span>affinity_m <span class="op">=</span> <span class="va">W</span>, label <span class="op">=</span> <span class="va">cell_label</span>, layout <span class="op">=</span> <span class="va">visual_space</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.key.height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"null"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/cell_graph-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="step2-diffusing-the-gene-expression-value-across-the-cell-graph">Step2: Diffusing the gene expression value across the cell
graph<a class="anchor" aria-label="anchor" href="#step2-diffusing-the-gene-expression-value-across-the-cell-graph"></a>
</h3>
<p>Next, we set the initial state of each gene by normalizing the
expression matrix using <code>RowwiseNormalize</code>. We then construct
a list of diffusion operators at different time scales using
<code>ConstructDiffusionOperators</code>. By multiplying the initial
state with the corresponding diffusion operators, we obtain the diffused
state of each gene at the dyadic time scales
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>0</mn><mo>,</mo><msup><mn>2</mn><mn>1</mn></msup><mo>,</mo><msup><mn>2</mn><mn>2</mn></msup><mo>,</mo><mi>…</mi></mrow><annotation encoding="application/x-tex">0,2^1,2^2,\dots</annotation></semantics></math>.
Finally, we can visualize the diffused state of each gene on the cell
embedding.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Construct a list of diffusion operators</span></span>
<span><span class="va">P_ls</span> <span class="op">=</span> <span class="fu"><a href="../reference/ConstructDiffusionOperators.html">ConstructDiffusionOperators</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">W</span>, max_time <span class="op">=</span> <span class="fl">2</span><span class="op">^</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt; Create a list of diffusion operators...</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |=======                                                               |  10%  |                                                                              |=====================                                                 |  30%  |                                                                              |============================                                          |  40%  |                                                                              |===================================                                   |  50%  |                                                                              |==========================================                            |  60%  |                                                                              |=================================================                     |  70%  |                                                                              |========================================================              |  80%  |                                                                              |===============================================================       |  90%  |                                                                              |======================================================================| 100%</span></span>
<span><span class="co">#&gt; Converting diffusion operators to sparse matrices...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Max diffusion time: 16384</span></span>
<span><span class="co"># Generate initial state for each gene</span></span>
<span><span class="va">rho</span> <span class="op">=</span> <span class="fu"><a href="../reference/RowwiseNormalize.html">RowwiseNormalize</a></span><span class="op">(</span><span class="va">dat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">W</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Here we visualize the diffused state of <em>Fcnb</em> - a marker for
granulocytopoietic cells, at four different time points:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mo>=</mo><mn>0</mn><mo>,</mo><mi>t</mi><mo>=</mo><msup><mn>2</mn><mn>1</mn></msup><mo>,</mo><mi>t</mi><mo>=</mo><msup><mn>2</mn><mn>4</mn></msup><mo>,</mo></mrow><annotation encoding="application/x-tex">t=0,t=2^1,t=2^4,</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mo>=</mo><msup><mn>2</mn><mn>10</mn></msup></mrow><annotation encoding="application/x-tex">t=2^{10}</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene</span> <span class="op">=</span> <span class="st">"Fcnb"</span></span>
<span><span class="fu"><a href="../reference/VisualizeDiffusion.html">VisualizeDiffusion</a></span><span class="op">(</span>coord <span class="op">=</span> <span class="va">visual_space</span>,init_state <span class="op">=</span> <span class="va">rho</span><span class="op">[</span><span class="va">gene</span>,<span class="op">]</span>,P_ls <span class="op">=</span> <span class="va">P_ls</span>,check_time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">^</span><span class="fl">4</span>,<span class="fl">2</span><span class="op">^</span><span class="fl">10</span><span class="op">)</span>,gene_name <span class="op">=</span> <span class="va">gene</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>  plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  plot.subtitle <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/gene_diffusion-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="step3-obtain-diffusion-kl-score-lmd-score">Step3: Obtain Diffusion KL score &amp; LMD score<a class="anchor" aria-label="anchor" href="#step3-obtain-diffusion-kl-score-lmd-score"></a>
</h3>
<p>We measure the dynamics of the diffusion process for each gene using
a score profile. This profile is calculated by the Kullback–Leibler (KL)
divergence between the initial state and the diffused state, and then
normalizing it by the KL divergence between the initial state and the
equilibrium state.</p>
<p>We summarize this score profile into the LMD score, which is the
cumulative sum of the score profile at all dyadic time steps. This LMD
score can be used to rank the genes. Both the score profile and the LMD
score are calculated using <code>fast_get_lmds</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">score_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/fast_get_lmds.html">fast_get_lmds</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">W</span>, init_state <span class="op">=</span> <span class="va">rho</span>, P_ls <span class="op">=</span> <span class="va">P_ls</span>, largeData <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Now, we visualize three sets of genes expressing in different number
of cells:</p>
<ul>
<li><p>Expressed in approximately
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>20</mn><annotation encoding="application/x-tex">20</annotation></semantics></math>
cells: <em>Tlr11</em>, <em>4933425H06Rik</em></p></li>
<li><p>Expressed in approximately
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>500</mn><annotation encoding="application/x-tex">500</annotation></semantics></math>
cells: <em>Fcnb</em>, <em>Uchl4</em></p></li>
<li><p>Expressed in approximately
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><mo>,</mo><mn>500</mn></mrow><annotation encoding="application/x-tex">2,500</annotation></semantics></math>
cells: <em>Cd79a</em>, <em>Dock11</em></p></li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Tlr11"</span>,<span class="st">"Fcnb"</span>,<span class="st">"Cd79a"</span>,<span class="st">"4933425H06Rik"</span>,<span class="st">"Uchl4"</span>,<span class="st">"Dock11"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">tiss</span>, features <span class="op">=</span> <span class="va">genes</span>, ncol <span class="op">=</span> <span class="fl">3</span>, order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>, legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>We visualize the score profiles for genes above. Genes with more
localized patterns (<em>Tlr11</em>, <em>Fcnb</em>, <em>Cd79a</em>) tend
to have smaller LMD scores (area under the curve) regardless of the
number of cells in which they are expressed.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"localized"</span>,<span class="st">"non_localized"</span><span class="op">)</span>,each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/VisualizeScorePattern.html">VisualizeScorePattern</a></span><span class="op">(</span><span class="va">score_result</span><span class="op">$</span><span class="st">'score_profile'</span>, genes <span class="op">=</span> <span class="va">genes</span>, label_class <span class="op">=</span> <span class="va">genes_label</span>, facet_class <span class="op">=</span> <span class="cn">NULL</span>, text <span class="op">=</span> <span class="cn">TRUE</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    plot.subtitle <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>, axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>, axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.7</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/score_profile-1.png" width="576"></p>
<p>The results of the LMD can be presented in a table, displaying the
LMD score and the rank for each gene.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/show_result_lmd.html">show_result_lmd</a></span><span class="op">(</span><span class="va">score_result</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">gene_table</span>,<span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  score rank</span></span>
<span><span class="co">#&gt; Mmp9          4.508234    1</span></span>
<span><span class="co">#&gt; 1100001G20Rik 4.624757    2</span></span>
<span><span class="co">#&gt; Itgb2l        4.627945    3</span></span>
<span><span class="co">#&gt; Lrg1          4.654115    4</span></span>
<span><span class="co">#&gt; C5ar1         4.737228    5</span></span>
<span><span class="co">#&gt; Retnlg        4.799311    6</span></span>
<span><span class="co">#&gt; Hp            4.799382    7</span></span>
<span><span class="co">#&gt; 1810033B17Rik 4.813253    8</span></span>
<span><span class="co">#&gt; Lcn2          4.825797    9</span></span>
<span><span class="co">#&gt; Chi3l1        4.828846   10</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-lmd-in-one-step">Running LMD in one step<a class="anchor" aria-label="anchor" href="#running-lmd-in-one-step"></a>
</h2>
<p>LMD can be run in one step. In this case, the input is:</p>
<ul>
<li><p><code>dat</code>: the log-normalized gene expression data, where
rows correspond to genes and columns correspond to cells.</p></li>
<li><p><code>feature_space</code>: the first 20 principal components
(PCs) of the <code>dat</code>.</p></li>
<li><p><code>knn</code>:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
of the cell-cell kNN graph.</p></li>
</ul>
<p>This example dataset contains 5,037 cells, running <code>LMD</code>
should take 1 to 3 minutes to finish.</p>
<p><strong>Note</strong>: The actual runtime can vary depending on your
system’s matrix operation performance. To potentially reduce runtime and
enhance performance, ensure that your R environment is configured with
optimized BLAS/LAPACK libraries, such as OpenBLAS, which support
parallel processing. For details on the specific setup used in this
tutorial, please refer to <a href="#session-information">Session
information</a>.</p>
<p>For more information on installing and configuring these libraries,
visit <a href="https://www.openblas.net/" class="external-link">OpenBLAS</a> and <a href="https://cran.r-project.org/doc/manuals/r-release/R-admin.html#OpenBLAS-and-BLIS" class="external-link">R
Administration and Installation</a>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://KlugerLab.github.io/LocalizedMarkerDetector/" class="external-link">LocalizedMarkerDetector</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jabiru/tictoc" class="external-link">tictoc</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">score_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/LMD.html">LMD</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">feature_space</span>, knn <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Constructing KNN graph</span></span>
<span><span class="co">#&gt; Remove 0 genes which express in less than 5 cells</span></span>
<span><span class="co">#&gt; Calculate LMD score profile for large data...</span></span>
<span><span class="co">#&gt; Run doubly stochastic on affinity matrix...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; max diffusion time:2^ 14</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 92.965 sec elapsed</span></span></code></pre></div>
<p>Present the results of LMD in a table: the LMD score and the rank for
each gene.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/show_result_lmd.html">show_result_lmd</a></span><span class="op">(</span><span class="va">score_result</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">gene_table</span>,<span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  score rank</span></span>
<span><span class="co">#&gt; Mmp9          4.508234    1</span></span>
<span><span class="co">#&gt; 1100001G20Rik 4.624757    2</span></span>
<span><span class="co">#&gt; Itgb2l        4.627945    3</span></span>
<span><span class="co">#&gt; Lrg1          4.654115    4</span></span>
<span><span class="co">#&gt; C5ar1         4.737228    5</span></span>
<span><span class="co">#&gt; Retnlg        4.799311    6</span></span>
<span><span class="co">#&gt; Hp            4.799382    7</span></span>
<span><span class="co">#&gt; 1810033B17Rik 4.813253    8</span></span>
<span><span class="co">#&gt; Lcn2          4.825797    9</span></span>
<span><span class="co">#&gt; Chi3l1        4.828846   10</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="improve-the-computational-efficiency-by-cell-graph-coarse-graining-">Improve the computational efficiency by cell graph
coarse-graining.<a class="anchor" aria-label="anchor" href="#improve-the-computational-efficiency-by-cell-graph-coarse-graining-"></a>
</h2>
<p>For large datasets, we recommend an optional cell graph
coarse-graining strategy adapted from <a href="https://klugerlab.github.io/GeneTrajectory/articles/fast_computation.html" class="external-link">GeneTrajectory</a><span class="citation">(Qu et al. 2024)</span>. to further reduce
computational cost. This approach aggregates cells into a smaller number
of “meta-cells” using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>-means
clustering, and defines the expression profile of each meta-cell by
summing the expression values of its constituent cells.</p>
<p>Below is an example of coarse-grain the cell graph by grouping cells
into N=1000 “meta-cells”.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg_output</span> <span class="op">=</span> <span class="fu"><a href="../reference/CoarseGrain.html">CoarseGrain</a></span><span class="op">(</span><span class="va">feature_space</span>, expression <span class="op">=</span> <span class="va">dat</span>, graph.affinity <span class="op">=</span> <span class="va">W</span>, N <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Run k-means clustering</span></span>
<span><span class="co">#&gt; Coarse-grain matrices</span></span>
<span><span class="va">W_cg</span> <span class="op">=</span> <span class="va">cg_output</span><span class="op">$</span><span class="va">graph</span></span>
<span><span class="va">dat_cg</span> <span class="op">=</span> <span class="va">cg_output</span><span class="op">$</span><span class="va">expression</span></span>
<span>    </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">score_result_cg</span> <span class="op">=</span> <span class="fu"><a href="../reference/LMD.html">LMD</a></span><span class="op">(</span><span class="va">dat_cg</span>, max_time <span class="op">=</span> <span class="fl">2</span><span class="op">^</span><span class="fl">20</span>, graph.affinity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">W_cg</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Remove 0 genes which express in less than 5 cells</span></span>
<span><span class="co">#&gt; Calculate LMD score profile for large data...</span></span>
<span><span class="co">#&gt; Run doubly stochastic on affinity matrix...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; max diffusion time:2^ 13</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 9.456 sec elapsed</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-gene-modules">Identifying Gene Modules<a class="anchor" aria-label="anchor" href="#identifying-gene-modules"></a>
</h2>
<p>First, we select the top localized genes based on the knee point of
the LMDS distribution.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/show_result_lmd.html">show_result_lmd</a></span><span class="op">(</span><span class="va">score_result</span>, kneeplot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/unnamed-chunk-15-1.png" width="480"></p>
<pre><code><span><span class="co">#&gt; knee_point:  1741</span></span>
<span><span class="va">top_lmd_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">cut_off_gene</span><span class="op">)</span></span></code></pre>
<p>Next, we compute the gene-gene pairwise Jaccard distance based on
their expression levels over all cells using
<code>CalculateGeneDistance</code>. We recommend users to apply <a href="https://github.com/KlugerLab/ALRA/tree/master" class="external-link">ALRA</a>
imputation<span class="citation">(Linderman et al. 2022)</span> first to
reduce the effect of drop-out events on the gene distance
calculations.</p>
<p><strong>Notes</strong>:</p>
<ol style="list-style-type: decimal">
<li><p>ALRA may produce different results with each run, leading to
slight variations in the gene-gene distance and gene clustering results
For reproducibility with this tutorial, users can download our <a href="https://figshare.com/ndownloader/files/48627340" class="external-link">preprocessed
Seurat object</a>, which includes the ALRA assay.</p></li>
<li><p>This step is time-consuming and may take 4 to 5 minutes to
complete.</p></li>
</ol>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># tic()</span></span>
<span><span class="co"># ALRA imputation</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"alra"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">@</span><span class="va">assays</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">tiss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratWrappers/man/RunALRA.html" class="external-link">RunALRA</a></span><span class="op">(</span><span class="va">tiss</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="op">}</span></span>
<span><span class="va">dat_alra</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">[[</span><span class="st">"alra"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="va">top_lmd_genes</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Compute the gene-gene pairwise distance</span></span>
<span><span class="va">dist</span> <span class="op">=</span> <span class="fu"><a href="../reference/CalculateGeneDistance.html">CalculateGeneDistance</a></span><span class="op">(</span><span class="va">dat_alra</span>, method <span class="op">=</span> <span class="st">"jaccard"</span><span class="op">)</span></span>
<span><span class="co"># toc()</span></span></code></pre></div>
<p>After obtaining the gene-gene distance matrix <code>dist</code>, we
cluster the genes using <code><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">stats::hclust</a></code> with
<code>average</code> option and determine the gene modules using
<code><a href="https://rdrr.io/pkg/dynamicTreeCut/man/cutreeDynamic.html" class="external-link">dynamicTreeCut::cutreeDynamic</a></code>. Optionally, we remove
outlier genes in each module with SML algorithm<span class="citation">(Parisi et al. 2014)</span> and discard modules
containing fewer than 10 genes. All of these steps are performed using
<code>ClusterGenes</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_cl_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/ClusterGenes.html">ClusterGenes</a></span><span class="op">(</span><span class="va">dist</span>, clustering_method <span class="op">=</span> <span class="st">"average"</span>, return_tree <span class="op">=</span> <span class="cn">TRUE</span>, deepSplit <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;  ..cutHeight not given, setting it to 0.976  ===&gt;  99% of the (truncated) height range in dendro.</span></span>
<span><span class="co">#&gt;  ..done.</span></span>
<span><span class="co">#&gt; Filtering out outlier genes in each module:  1684  genes left.</span></span></code></pre></div>
<p>We can visualize the pairwise distance matrix of genes. The sidebar
and side tree represent the partitioning of genes into gene modules.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeGeneHeatmap.html">VisualizeGeneHeatmap</a></span><span class="op">(</span><span class="va">dist</span>, gene_partition <span class="op">=</span> <span class="va">gene_cl_res</span><span class="op">$</span><span class="va">gene_partition</span>, gene_hree <span class="op">=</span> <span class="va">gene_cl_res</span><span class="op">$</span><span class="va">gene_hree</span><span class="op">)</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/unnamed-chunk-18-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="computing-per-cell-module-activity-scores">Computing Per-Cell Module Activity Scores<a class="anchor" aria-label="anchor" href="#computing-per-cell-module-activity-scores"></a>
</h2>
<p>Next, we calculate the module activity score for each module using
<code>AddModuleActivityScore</code>, which represents the probability of
a module being expressed in a given cell. This score can be further used
to visualize the expression patterns of each module and to identify the
corresponding cells.</p>
<p><strong>Note</strong>: This step is CPU-intensive and may take 4 to 5
minutes to complete. We also suggest considering alternative methods,
such as <a href="https://satijalab.org/seurat/reference/addmodulescore" class="external-link"><code>Seurat::AddModuleScore</code></a>,
to visualize the gene module on cell embedding.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_partition</span> <span class="op">=</span> <span class="va">gene_cl_res</span><span class="op">$</span><span class="va">gene_partition</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">gene_partition</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nlevels.html" class="external-link">nlevels</a></span><span class="op">(</span><span class="va">gene_partition</span><span class="op">)</span> <span class="co"># rename gene modules</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tiss</span> <span class="op">=</span> <span class="fu"><a href="../reference/AddModuleActivityScore.html">AddModuleActivityScore</a></span><span class="op">(</span><span class="va">tiss</span>, gene_partition <span class="op">=</span> <span class="va">gene_partition</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize patterns</span></span>
<span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">tiss</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Module"</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tiss</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>, order <span class="op">=</span> <span class="cn">TRUE</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"lightgrey"</span>, high <span class="op">=</span> <span class="st">"blue"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ModuleScore"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/gene_module-1.png" width="960"></p>
<p>We can take a closer look at a module by examining several of its top
localized genes.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">module_id</span> <span class="op">=</span> <span class="fl">19</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">tiss</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Module"</span>,<span class="va">module_id</span><span class="op">)</span>, order <span class="op">=</span> <span class="cn">TRUE</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Module%s (%d genes)"</span>,<span class="va">module_id</span>,<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">gene_partition</span> <span class="op">==</span> <span class="va">module_id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ModuleScore"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">pl</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">tiss</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_partition</span><span class="op">)</span><span class="op">[</span><span class="va">gene_partition</span> <span class="op">==</span> <span class="va">module_id</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>,ncol <span class="op">=</span> <span class="fl">3</span>, order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">=</span> <span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">pl</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#BBB\nABBB\nABBB\n#BBB"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="LMD_demo_files/figure-html/gene_module_genes-1.png" width="960"></p>
<p>To better understand the biological functions of these gene modules,
we can perform a series of functional enrichment analyses, such as Gene
Ontology (GO) enrichment analysis<span class="citation">(Wu et al.
2021)</span> or Reactome Pathway Database enrichment analysis<span class="citation">(Yu and He 2016)</span>. Here, we demonstrate the
Reactome Pathway Database enrichment analysis using
<code><a href="https://rdrr.io/pkg/ReactomePA/man/enrichPathway.html" class="external-link">ReactomePA::enrichPathway</a></code> for previous selected gene
module, and we print its top 5 significantly enriched pathways.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">ReactomePA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bg_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span> <span class="co"># Set all genes(after filtering) in the expression matrix as background genes</span></span>
<span><span class="va">universe_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"symbol"</span> <span class="op">=</span> <span class="va">bg_genes</span>,<span class="st">"entrez"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Mm.eg.db</span>, keys<span class="op">=</span><span class="va">bg_genes</span>, column<span class="op">=</span><span class="st">"ENTREZID"</span>, keytype<span class="op">=</span><span class="st">"SYMBOL"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">universe_df</span> <span class="op">=</span> <span class="va">universe_df</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">universe_df</span><span class="op">$</span><span class="va">entrez</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">epathway_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReactomePA/man/enrichPathway.html" class="external-link">enrichPathway</a></span><span class="op">(</span>gene<span class="op">=</span><span class="va">universe_df</span><span class="op">[</span><span class="va">universe_df</span><span class="op">$</span><span class="va">symbol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_partition</span><span class="op">)</span><span class="op">[</span><span class="va">gene_partition</span> <span class="op">==</span> <span class="va">module_id</span><span class="op">]</span>,<span class="st">"entrez"</span><span class="op">]</span>,</span>
<span>                organism <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>                pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                universe <span class="op">=</span> <span class="va">universe_df</span><span class="op">$</span><span class="va">entrez</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">epathway_result</span><span class="op">@</span><span class="va">result</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Description'</span>, <span class="st">'GeneRatio'</span>, <span class="st">'p.adjust'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                          Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; R-MMU-1640170             Cell Cycle     30/47 3.388436e-23</span></span>
<span><span class="co">#&gt; R-MMU-69278      Cell Cycle, Mitotic     28/47 7.167466e-23</span></span>
<span><span class="co">#&gt; R-MMU-68886                  M Phase     21/47 6.462027e-17</span></span>
<span><span class="co">#&gt; R-MMU-68877     Mitotic Prometaphase     18/47 1.268621e-16</span></span>
<span><span class="co">#&gt; R-MMU-69620   Cell Cycle Checkpoints     18/47 1.228419e-15</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.1.3 (2022-03-10)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] org.Mm.eg.db_3.14.0           AnnotationDbi_1.56.2         </span></span>
<span><span class="co">#&gt;  [3] IRanges_2.28.0                S4Vectors_0.32.4             </span></span>
<span><span class="co">#&gt;  [5] Biobase_2.54.0                BiocGenerics_0.40.0          </span></span>
<span><span class="co">#&gt;  [7] ReactomePA_1.38.0             patchwork_1.1.2              </span></span>
<span><span class="co">#&gt;  [9] ggplot2_3.3.6                 tictoc_1.2                   </span></span>
<span><span class="co">#&gt; [11] SeuratWrappers_0.3.1          SeuratObject_4.1.3           </span></span>
<span><span class="co">#&gt; [13] Seurat_4.3.0.1                LocalizedMarkerDetector_1.0.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] rappdirs_0.3.3          scattermore_0.8         R.methodsS3_1.8.2      </span></span>
<span><span class="co">#&gt;   [4] ragg_1.3.3              tidyr_1.2.0             bit64_4.0.5            </span></span>
<span><span class="co">#&gt;   [7] knitr_1.49              irlba_2.3.5             R.utils_2.12.0         </span></span>
<span><span class="co">#&gt;  [10] data.table_1.14.2       KEGGREST_1.34.0         RCurl_1.98-1.8         </span></span>
<span><span class="co">#&gt;  [13] generics_0.1.3          cowplot_1.1.1           RSQLite_2.2.16         </span></span>
<span><span class="co">#&gt;  [16] shadowtext_0.1.2        RANN_2.6.1              future_1.27.0          </span></span>
<span><span class="co">#&gt;  [19] bit_4.0.4               enrichplot_1.14.2       spatstat.data_3.0-1    </span></span>
<span><span class="co">#&gt;  [22] httpuv_1.6.5            viridis_0.6.2           xfun_0.51              </span></span>
<span><span class="co">#&gt;  [25] hms_1.1.2               jquerylib_0.1.4         evaluate_1.0.3         </span></span>
<span><span class="co">#&gt;  [28] promises_1.2.0.1        additivityTests_1.1-4.2 fansi_1.0.3            </span></span>
<span><span class="co">#&gt;  [31] progress_1.2.2          igraph_1.3.4            DBI_1.1.3              </span></span>
<span><span class="co">#&gt;  [34] htmlwidgets_1.5.4       spatstat.geom_3.2-5     purrr_1.0.2            </span></span>
<span><span class="co">#&gt;  [37] ellipsis_0.3.2          RSpectra_0.16-1         dplyr_1.1.3            </span></span>
<span><span class="co">#&gt;  [40] backports_1.4.1         deldir_1.0-6            vctrs_0.6.3            </span></span>
<span><span class="co">#&gt;  [43] remotes_2.4.2           ROCR_1.0-11             abind_1.4-5            </span></span>
<span><span class="co">#&gt;  [46] cachem_1.0.6            withr_3.0.2             ggforce_0.4.1          </span></span>
<span><span class="co">#&gt;  [49] progressr_0.10.1        checkmate_2.1.0         sctransform_0.3.5      </span></span>
<span><span class="co">#&gt;  [52] treeio_1.18.1           prettyunits_1.1.1       goftest_1.2-3          </span></span>
<span><span class="co">#&gt;  [55] cluster_2.1.2           DOSE_3.20.1             ape_5.6-2              </span></span>
<span><span class="co">#&gt;  [58] lazyeval_0.2.2          crayon_1.5.1            spatstat.explore_3.2-1 </span></span>
<span><span class="co">#&gt;  [61] pkgconfig_2.0.3         labeling_0.4.2          tweenr_2.0.2           </span></span>
<span><span class="co">#&gt;  [64] GenomeInfoDb_1.30.1     nlme_3.1-155            rlang_1.1.1            </span></span>
<span><span class="co">#&gt;  [67] globals_0.16.0          lifecycle_1.0.3         miniUI_0.1.1.1         </span></span>
<span><span class="co">#&gt;  [70] dbscan_1.1-12           rsvd_1.0.5              polyclip_1.10-0        </span></span>
<span><span class="co">#&gt;  [73] matrixStats_0.62.0      lmtest_0.9-40           graph_1.72.0           </span></span>
<span><span class="co">#&gt;  [76] Matrix_1.6-1            aplot_0.2.1             zoo_1.8-10             </span></span>
<span><span class="co">#&gt;  [79] ggridges_0.5.3          pheatmap_1.0.12         png_0.1-7              </span></span>
<span><span class="co">#&gt;  [82] viridisLite_0.4.1       bitops_1.0-7            R.oo_1.25.0            </span></span>
<span><span class="co">#&gt;  [85] KernSmooth_2.23-20      Biostrings_2.62.0       blob_1.2.3             </span></span>
<span><span class="co">#&gt;  [88] stringr_1.4.1           qvalue_2.26.0           parallelly_1.32.1      </span></span>
<span><span class="co">#&gt;  [91] spatstat.random_3.1-5   gridGraphics_0.5-1      reactome.db_1.77.0     </span></span>
<span><span class="co">#&gt;  [94] scales_1.2.1            memoise_2.0.1           graphite_1.40.0        </span></span>
<span><span class="co">#&gt;  [97] magrittr_2.0.3          plyr_1.8.7              ica_1.0-3              </span></span>
<span><span class="co">#&gt; [100] zlibbioc_1.40.0         compiler_4.1.3          scatterpie_0.2.1       </span></span>
<span><span class="co">#&gt; [103] RColorBrewer_1.1-3      fitdistrplus_1.1-8      cli_3.6.1              </span></span>
<span><span class="co">#&gt; [106] XVector_0.34.0          listenv_0.8.0           pbapply_1.5-0          </span></span>
<span><span class="co">#&gt; [109] MASS_7.3-55             tidyselect_1.2.0        stringi_1.7.8          </span></span>
<span><span class="co">#&gt; [112] textshaping_1.0.0       yaml_2.3.10             GOSemSim_2.20.0        </span></span>
<span><span class="co">#&gt; [115] ggrepel_0.9.1           grid_4.1.3              sass_0.4.9             </span></span>
<span><span class="co">#&gt; [118] fastmatch_1.1-3         tools_4.1.3             future.apply_1.9.0     </span></span>
<span><span class="co">#&gt; [121] parallel_4.1.3          rstudioapi_0.14         gridExtra_2.3          </span></span>
<span><span class="co">#&gt; [124] farver_2.1.1            Rtsne_0.16              ggraph_2.1.0           </span></span>
<span><span class="co">#&gt; [127] digest_0.6.37           BiocManager_1.30.18     FNN_1.1.3.1            </span></span>
<span><span class="co">#&gt; [130] shiny_1.7.2             flexclust_1.5.0         Rcpp_1.0.9             </span></span>
<span><span class="co">#&gt; [133] later_1.3.0             RcppAnnoy_0.0.19        httr_1.4.4             </span></span>
<span><span class="co">#&gt; [136] colorspace_2.0-3        biclust_2.0.3.1         fs_1.5.2               </span></span>
<span><span class="co">#&gt; [139] tensor_1.5              reticulate_1.25         splines_4.1.3          </span></span>
<span><span class="co">#&gt; [142] uwot_0.1.14             yulab.utils_0.0.9       tidytree_0.4.5         </span></span>
<span><span class="co">#&gt; [145] spatstat.utils_3.1-2    pkgdown_2.1.1           latex2exp_0.9.6        </span></span>
<span><span class="co">#&gt; [148] graphlayouts_1.0.0      sp_1.5-0                ggplotify_0.1.2        </span></span>
<span><span class="co">#&gt; [151] plotly_4.10.0           systemfonts_1.2.1       xtable_1.8-4           </span></span>
<span><span class="co">#&gt; [154] jsonlite_1.9.1          ggtree_3.2.1            dynamicTreeCut_1.63-1  </span></span>
<span><span class="co">#&gt; [157] tidygraph_1.2.3         modeltools_0.2-23       ggfun_0.1.3            </span></span>
<span><span class="co">#&gt; [160] R6_2.5.1                pillar_1.9.0            htmltools_0.5.8.1      </span></span>
<span><span class="co">#&gt; [163] mime_0.12               glue_1.6.2              fastmap_1.2.0          </span></span>
<span><span class="co">#&gt; [166] BiocParallel_1.28.3     class_7.3-20            codetools_0.2-18       </span></span>
<span><span class="co">#&gt; [169] fgsea_1.20.0            utf8_1.2.2              lattice_0.20-45        </span></span>
<span><span class="co">#&gt; [172] bslib_0.9.0             spatstat.sparse_3.0-2   tibble_3.2.1           </span></span>
<span><span class="co">#&gt; [175] leiden_0.4.2            philentropy_0.7.0       GO.db_3.14.0           </span></span>
<span><span class="co">#&gt; [178] survival_3.2-13         rmarkdown_2.29          desc_1.4.3             </span></span>
<span><span class="co">#&gt; [181] munsell_0.5.0           DO.db_2.9               GenomeInfoDbData_1.2.7 </span></span>
<span><span class="co">#&gt; [184] reshape2_1.4.4          gtable_0.3.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-tabula2018single" class="csl-entry">
Consortium, Tabula Muris, Overall coordination Schaum Nicholas 1
Karkanias Jim 2 Neff Norma F. 2 May Andrew P. 2 Quake Stephen R. quake@
stanford. edu 2 3 f Wyss-Coray Tony twc@ stanford. edu 4 5 6 g Darmanis
Spyros spyros. darmanis@ czbiohub. org 2 h, Logistical coordination
Batson Joshua 2 Botvinnik Olga 2 Chen Michelle B. 3 Chen Steven 2 Green
Foad 2 Jones Robert C. 3 Maynard Ashley 2 Penland Lolita 2 Pisco Angela
Oliveira 2 Sit Rene V. 2 Stanley Geoffrey M. 3 Webber James T. 2 Zanini
Fabio 3, and Computational data analysis Batson Joshua 2 Botvinnik Olga
2 Castro Paola 2 Croote Derek 3 Darmanis Spyros 2 DeRisi Joseph L. 2 27
Karkanias Jim 2 Pisco Angela Oliveira 2 Stanley Geoffrey M. 3 Webber
James T. 2 Zanini Fabio 3. 2018. <span>“Single-Cell Transcriptomics of
20 Mouse Organs Creates a Tabula Muris.”</span> <em>Nature</em> 562
(7727): 367–72.
</div>
<div id="ref-linderman2022zero" class="csl-entry">
Linderman, George C, Jun Zhao, Manolis Roulis, Piotr Bielecki, Richard A
Flavell, Boaz Nadler, and Yuval Kluger. 2022. <span>“Zero-Preserving
Imputation of Single-Cell RNA-Seq Data.”</span> <em>Nature
Communications</em> 13 (1): 192.
</div>
<div id="ref-parisi2014ranking" class="csl-entry">
Parisi, Fabio, Francesco Strino, Boaz Nadler, and Yuval Kluger. 2014.
<span>“Ranking and Combining Multiple Predictors Without Labeled
Data.”</span> <em>Proceedings of the National Academy of Sciences</em>
111 (4): 1253–58.
</div>
<div id="ref-qu2024gene" class="csl-entry">
Qu, Rihao, Xiuyuan Cheng, Esen Sefik, Jay S Stanley III, Boris Landa,
Francesco Strino, Sarah Platt, et al. 2024. <span>“Gene Trajectory
Inference for Single-Cell Data by Optimal Transport Metrics.”</span>
<em>Nature Biotechnology</em>, 1–11.
</div>
<div id="ref-wu2021clusterprofiler" class="csl-entry">
Wu, Tianzhi, Erqiang Hu, Shuangbin Xu, Meijun Chen, Pingfan Guo, Zehan
Dai, Tingze Feng, et al. 2021. <span>“clusterProfiler 4.0: A Universal
Enrichment Tool for Interpreting Omics Data.”</span> <em>The
Innovation</em> 2 (3).
</div>
<div id="ref-yu2016reactomepa" class="csl-entry">
Yu, Guangchuang, and Qing-Yu He. 2016. <span>“ReactomePA: An
r/Bioconductor Package for Reactome Pathway Analysis and
Visualization.”</span> <em>Molecular BioSystems</em> 12 (2): 477–79.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ruiqi Li, Rihao Qu, Francesco Strino.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
